Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -314,9 +378,260 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+def nonEmojis : Array Char := #[]
+-- TODO: should there be any? maybe #['↗', '↘', '✝', '▼', '▶']?
+
+/--
+Other unicode characters present in Mathlib (as of Aug. 28, 2024)
+not present in any of the above lists.
+-/
+def othersInMathlib := #[
+'▼', 'c', 'ō', '⏩', '❓',
+'🆕', 'š', 'ř', '⚬', '│', '├', '┌', 'ő',
+'⟍', '̂', 'ᘁ', 'ń', 'ć', '⟋', 'ỳ', 'ầ', '⥥',
+'ł', '◿', '◹', '－', '＼', '◥', '／', '◢', 'Ž', 'ă',
+'И', 'в', 'а', 'н', 'о', 'и', 'ч', 'Š', 'ᴜ', 'ᵧ', '´',
+'ᴄ', 'ꜰ', 'ß', 'ᴢ', 'ᴏ', 'ᴀ', 'ꜱ', 'ɴ', 'ꞯ', 'ʟ',
+'ʜ', 'ᵟ', 'ʙ', 'ᵪ', 'ᵩ', 'ᵦ', 'ᴊ', 'ᴛ', 'ᴡ', 'ᴠ',
+'ɪ', '̀', 'ᴇ', 'ᴍ', 'ʀ', 'ᴅ', 'ɢ', 'ʏ', 'ᴘ', 'ĝ', 'ᵨ',
+'ᴋ', 'ś', '꙳', '𝓡', '𝕝', '𝖣', '⨳' ]
+
+/-
+TODO there are more symbols we could use that aren't in this list yet. E.g, see
+https://en.wikipedia.org/wiki/Mathematical_operators_and_symbols_in_Unicode
+-/
+
+/-- If `false` the character is not allowed in Mathlib.
+Consider adding it to one of the whitelists. -/
+def isAllowedCharacter (c : Char) : Bool :=
+  ASCII.whitespace c
+  || ASCII.printable c
+  || withVSCodeAbbrev.contains c
+  || emojis.contains c
+  || nonEmojis.contains c
+  || c == UnicodeVariant.emoji
+  || c == UnicodeVariant.text
+  || othersInMathlib.contains c
+
+/-- Creates `StyleError`s for non-whitelisted unicode characters as well as
+bad usage of (emoji/text)-variant-selectors. -/
+def findBadUnicodeAux (s : String) (pos : String.Pos) (c : Char)
+    (err : Array StyleError := #[]) : Array StyleError :=
+  if h : pos < s.endPos then
+    have := Nat.sub_lt_sub_left h (String.lt_next s pos)
+    let posₙ := s.next pos
+    let cₙ := s.get? posₙ |>.getD '\uFFFD' -- �
+    if cₙ == UnicodeVariant.emoji && !(emojis.contains c) then
+      -- bad: unwanted emoji-variant-selector
+      let errₙ := err.push (.unicodeVariant ⟨[c, cₙ]⟩ none pos)
+      findBadUnicodeAux s posₙ cₙ errₙ
+    else if cₙ == UnicodeVariant.text && !(nonEmojis.contains c) then
+      -- bad: unwanted text-variant selector
+      let errₙ := err.push (.unicodeVariant ⟨[c, cₙ]⟩ none pos)
+      findBadUnicodeAux s posₙ cₙ errₙ
+    else if cₙ != UnicodeVariant.emoji && emojis.contains c then
+      -- bad: missing emoji-variant selector
+      let errₙ := err.push (.unicodeVariant ⟨[c, cₙ]⟩ UnicodeVariant.emoji pos)
+      findBadUnicodeAux s posₙ cₙ errₙ
+    else if cₙ != UnicodeVariant.text && nonEmojis.contains c then
+      -- bad: missing text-variant selector
+      let errₙ := err.push (.unicodeVariant ⟨[c, cₙ]⟩ UnicodeVariant.text pos)
+      findBadUnicodeAux s posₙ cₙ errₙ
+    else if ! isAllowedCharacter c then
+      -- bad: character not allowed
+      let errₙ := err.push (.unwantedUnicode c)
+      findBadUnicodeAux s posₙ cₙ errₙ
+    else
+      -- okay
+      findBadUnicodeAux s posₙ cₙ err
+  -- emojis/non-emojis should not be the last character in the line
+  -- in practise, the last char is `\n`, so this is in theory superfluous.
```

Comment:
I believe with the way this linter is called, the final \n is not part of the "line" we process.

Reply:
I removed the comment and kept the check. Please double-check if you agree.